lib/datadog/core/configuration/option.rb

```diff
@@ -6,20 +6,46 @@ module Configuration
       # Represents an instance of an integration configuration option
       # @public_api
       class Option
-        attr_reader \
-          :definition
+        attr_reader :definition
+
+        # Option setting precedence. Higher number means higher precedence.
+        module Precedence
+          REMOTE_CONFIGURATION = [2, :remote_configuration].freeze
+          PROGRAMMATIC = [1, :programmatic].freeze
+          DEFAULT = [0, :default].freeze
+        end
 
         def initialize(definition, context)
           @definition = definition
           @context = context
           @value = nil
           @is_set = false
+
+          # Lowest precedence, to allow for `#set` to always succeed for a brand new `Option` instance.
+          @precedence_set = Precedence::DEFAULT
         end
 
-        def set(value)
+        # Overrides the current value for this option if the `precedence` is equal or higher than
+        # the previously set value.
+        # The first call to `#set` will always store the value regardless of precedence.
+        #
+        # @param value [Object] the new value to be associated with this option
+        # @param precedence [Precedence] from what precedence order this new value comes from
+        def set(value, precedence: Precedence::PROGRAMMATIC)
+          # Cannot override higher precedence value
+          if precedence[0] < @precedence_set[0]
+            Datadog.logger.debug do
+              "Option '#{definition.name}' not changed to '#{value}' (precedence: #{precedence[1]}) because the higher " \
+                "precedence value '#{@value}' (precedence: #{@precedence_set[1]}) was already set."
+            end
+
+            return @value
```

Comment:
I was concerned with our own internal usage of [`Datadog.configure {}`](https://github.com/search?q=repo%3ADataDog%2Fdd-trace-rb+%2FDatadog.configure%5Cb%2F+path%3A%2F%5Elib%5C%2Fdatadog%5C%2Ftracing%5C%2Fcontrib%5C%2F%2F&type=code), which the user has no control over. 
These, in my mind, could trigger this log line.

But inspecting them closer, none of the internal use cases change any of the options that Remote Configuration change today, and they don't seem likely to be supported through Remote Configuration in the future.

I've bumped it up to INFO: I'm not confident WARN is the best option because Remote Configuration overriding your local configuration leaves the application in a stable state, so the WARN might not actually be of concern. I don't want to have to introduce an introspection API for users to avoid triggering the WARN ("let me check if this option was set by remote configuration before setting it locally, otherwise ddtrace will WARN").

Reply:
Seems reasonable!